### PR TITLE
Start making `k8s.io/client-go` authoritative for generic client packages

### PR DESCRIPTION
**Pull request by deads2k**
Right now, client-go has copies of various generic client packages which produces golang type incompatibilities when you want to switch between different kinds of clients.  In many cases, there's no reason to have two sets of packages.  This pull eliminates the copy for `pkg/client/transport` and makes `client-go` the authoritative copy.

I recommend going by commits, the first just synchronizes the client-go code again so that I could test the copy script to make sure it correctly preserves the original package.

@kubernetes/sig-api-machinery-misc @lavalamp @sttts

### CODE REVIEW

File: pkg/client/restclient/transport.go [hunk 1]
```diff
@@ -20,7 +20,7 @@ import (
 	"crypto/tls"
 	"net/http"
 
-	"k8s.io/kubernetes/pkg/client/transport"
+	"k8s.io/client-go/transport"
```
Comment (lavalamp) [inline pkg/client/restclient/transport.go]: This is a bit confusing-- the file you're editing is itself compiled into the client library. Is this really right?

File: staging/copy.sh [hunk 2]
```diff
@@ -60,6 +60,12 @@ function mkcp() {
     mkdir -p "${CLIENT_REPO_TEMP}/$2" && cp -r "${MAIN_REPO}/$1" "${CLIENT_REPO_TEMP}/$2"
 }
 
+# save copies code from client-go into the temp folder to make sure we don't lose it by accident
+# TODO this is temporary until everything in certain directories is authoritative
+function save() {
```
Comment (lavalamp) [inline staging/copy.sh]: I see. So you want to track some files in /staging as canonical, while some are still programmatically generated from elsewhere? This is going to be super confusing. Can you add an explanatory comment near the top of this file, that there are two classes of things in the staging directory?

File: staging/copy.sh [hunk 3]
```diff
@@ -70,7 +76,7 @@ mkcp "/pkg/client/restclient" "/pkg/client"
 mkcp "/pkg/client/testing" "/pkg/client"
 # remove this test because it imports the internal clientset
 rm "${CLIENT_REPO_TEMP}"/pkg/client/testing/core/fake_test.go
-mkcp "/pkg/client/transport" "/pkg/client"
+save "/transport"
```
Comment (lavalamp) [inline staging/copy.sh]: and I recommend splitting this into two sections--

```
# save everything for which the staging directory is the source of truth
save "/transport"
...

# assemble all the other parts of the staging directory
mvfolder ...
...
```

File: staging/copy.sh [hunk 4]
```diff
@@ -202,6 +207,9 @@ find "${CLIENT_REPO_TEMP}" -type f \( \
     -name "*.sh" \
     \) -delete
 
+echo "remove cyclical godep"
```
Comment (lavalamp) [inline staging/copy.sh]: this is a little scary.

File: staging/copy.sh (hunk 2 above)
Comment (deads2k) [inline staging/copy.sh]: > I see. So you want to track some files in /staging as canonical, while some are still programmatically generated from elsewhere? This is going to be super confusing. Can you add an explanatory comment near the top of this file, that there are two classes of things in the staging directory?

Ok.  I do see this as mostly intermediate. Eventually the staging copy is just API types and we simply regenerate the clients.

**Comment by lavalamp** [comment]
I have a few suggestions, since this is adding a new category of packages in the staging directory.

@krousey any thoughts?

File: staging/copy.sh (hunk 4 above)
Comment (deads2k) [inline staging/copy.sh]: > this is a little scary.

Yeah, but there's an entire separate go program devoted to managing the file, so I'm not the guy who discover weird godep behavior.  In all honesty, this whole, "copy, mutate, revendor, tweak vendor" which pre-exists my pull is pretty wild.

File: pkg/client/restclient/transport.go (hunk 1 above)
Comment (deads2k) [inline pkg/client/restclient/transport.go]: > This is a bit confusing-- the file you're editing is itself compiled into the client library. Is this really right?

Yes, this is where the package is used.

File: hack/make-rules/test.sh [hunk 5]
```diff
@@ -57,6 +57,10 @@ kube::test::find_dirs() {
         -o -path './test/e2e_node/system/*' \
       -name '*_test.go' -print0 | xargs -0n1 dirname | sed "s|^\./|${KUBE_GO_PACKAGE}/|" | LC_ALL=C sort -u
 
+    # run tests for client-go
+    find ./staging/src/k8s.io/client-go -name '*_test.go' \
```
Comment (sttts) [inline hack/make-rules/test.sh]: Does this mean we run a number of tests twice (until we have removed all code copying?). The unit test CI job takes long already now.

**Comment by sttts** [comment]
To put this PR into perspective: we will continue to make more code authorative in client-go? Are we going to remove the code copying altogether at some point?

I still have a number of question marks around composibility of the approach if we want clients for other, external apigroups (think of downstream projects). Will we end up with k8s.io/generic-client-go some day and make that authorative? Will the actual api types become authorative in the client-go package for a certain api group some day?

```
k8s.io/generic-client-go
k8s.io/core-client-go
k8s.io/service-catalog-client-go
github.com/openshift/origin-client-go
```

**Comment by deads2k** [comment]
> Will we end up with k8s.io/generic-client-go some day and make that authorative?

I think we should, but that's a big step to take right now.



> Will the actual api types become authorative in the client-go package for a certain api group some day?

No.  I don't think that ever happens.  I suspect they remain copied and pinned.



> we will continue to make more code authorative in client-go? Are we going to remove the code copying altogether at some point?

I think we can get it down to the point where you copy the API types and then generate the clients a second time.

File: hack/make-rules/test.sh (hunk 5 above)
Comment (deads2k) [inline hack/make-rules/test.sh]: > Does this mean we run a number of tests twice (until we have removed all code copying?). The unit test CI job takes long already now.

Yes.  Change hurts.

**Comment by sttts** [comment]
> I think we can get it down to the point where you copy the API types and then generate the clients a second time.

:+1:

**Comment by sttts** [comment]
lgtm from my side

**Comment by deads2k** [comment]
Comments addressed.

**Comment by lavalamp** [comment]
Just for the record I agree with @deads2k's responses.

**Comment by caesarxuchao** [comment]
> I think we can get it down to the point where you copy the API types and then generate the clients a second time.

I don't think we should copy the API types. We should separate the API types into their own repo and vendor them in client-go (and in other repos). Otherwise we will end up with different types in different repos, though their contents are the same.

**Comment by deads2k** [comment]
> I don't think we should copy the API types. We should separate the API types into their own repo and vendor them in client-go (and in other repos). Otherwise we will end up with different types in different repos, though their contents are the same.

This causes weird things to happen when you have a current level of API objects and you want to talk to an older API server.  Logically you want client-go, but because of the way vendored deps work, you end up flattening it down.  The `apis/v1` you wanted isn't the `apis/v1` that you'll get potentially leading to weird side-effects.  Imagine moving types between namespaced and not namespaced and changing supported verbs amongs others.

I think there is a reasonable argument for keeping them separate and copied.  @smarterclayton convinced me after I pitched doing the move.

**Pull request merged**